### DISCUSSION TITLE
Replace/Regex edit before importing

### DISCUSSION THREAD

**Discussion by JOJ0** [beetbox/beets#4571] [Ideas]
~This has been asked before: https://github.com/beetbox/beets/discussions/4047~

Update: Actually not the same question, I was wrong when I wrote that back then.

Sorry for repeating it but it's in a specific way I'm sure a lot of DJs and beatport users/customers in general will be aware of: 

Downloaded tracks from Beatport tend to look like this often: "Trackname (Original Mix)" - This will get beets to apply penalties when searching for tracknames on MusicBrainz, Discogs or whatever source plugin is in use. (except the source plugin is the beatport plugion itself)

https://github.com/edgars-supe/beets-importreplace does the opposite, not what I need.

It would help to just kick out "(Original Mix)" before searching online. Is there any solution existing already? The `fromfilename` plugin does something like that but it's not configurable. Would a PR making it possible to add new regex replacements to `fromfilename` be welcomed?

**Comment by benpage2** [comment]
Edit: This comment was from a different thread.

Not that I know of. I'm currently using mp3tag to make a custom `temp_title` field, regex replacing in mp3tag, importing without move with beets, then using mp3tag to re apply the `temp_title` field before using `beet move` to put everything in my library as is. 

Also a DJ and I'm really working hard to circumvent a lot of beets limitations because the lastgenre and whatlastgenre extensions are so good. If you have any workflows that work well for you please let me know lol

**Comment by arsaboo** [comment]
@JOJ0 I don't think there is one. We could extend the `importreplace` plugin to do some replacements before imports or incorporate that feature (regex replacement before and after imports) in beets. It is an important enough feature to be part of the core.

**Reply by benpage2** [reply]
Personally I don't see why not. It also seems like it would fit neatly into the `modify` command as well just for general use. Or maybe even its own `replace` command?

**Reply by JOJ0** [reply]
We have quite some plugins that replace something and have that in their names already. Let's carefully think about where to put that kind of functionality to not confuse the beets replacement jungle further 😅

**Reply by benpage2** [reply]
I only know of import-replace. Is there one that allows you to do regex replacements on your library without importing?

**Reply by JOJ0** [reply]
https://beets.readthedocs.io/en/latest/plugins

Ok replace not so much in the names, but similar terms like rewrite, substitute 

Where would that functionality fit best you think? And how could the config entry look like?

**Comment by benpage2** [comment]
So I've been thinking about what exactly would be most helpful for the last little bit @JOJ0 @arsaboo.

For sure the best thing would be a simple find and replace command with the ability to use regex. The idea being that it would act very similarly to `beet modify` and act on the library and tags. It may compliment the [import-replace plugin](https://github.com/edgars-supe/beets-importreplace)

`beet replace [options] QUERY find:replace`

Similar to `QUERY` `find:replace` runs a simple find and replace while `find::replace` runs a regex find and replace
Running `beet replace [options]` would run the replacements set under in the config (thinking that the config is similar to import-replace) but without having to do an import.

Options (same as `beet modify`):
- `-M` don't move
- `-W` don't write 
- `-w` write (default)
- `-a` albums instead of items
- `-y` skip yes prompt

Additional options:
- `-i` case insensitive
- `-e` matches entire word (using -e because "whole word" would intuitively be -w)
- `-f` specify a field? I'm not sure if this could be part of the query functionality or should be it's own option

Config options:
- `auto: yes/no` (default yes) run config on import
- `case_insensitive: yes/no` (default no)
- `entire_word: yes/no` (default no)

Additionally the config options from import-replace would be included (quoting from the import-replace readme):

- `item_fields` is a set of fields to run replacements on for items
- `album_fields` is a set of fields to run replacements on for albums
- `replace` is a set of regex/replacement pairs

Overall the config may look like:
```
pluginname:
    auto: yes
    case_insensitive: no
    entire_word: no
    replacements:
        - item_fields: title
           replace:
               'The':'That' (alternatively `The::'that' would be regex)
        - item_fields: album
        - album_fields: album
            replace:
                'Foo':'Bar'
```

**Comment by arsaboo** [comment]
@benpage2 This looks great. We can finetune the options as we start working on it. Once @JOJ0 gives their blessing, we can get started with this.

**Reply by benpage2** [reply]
Thanks! 

Just a heads up, I'm (unfortunately) a student in between exams and a new summer job at the moment. I have a fair bit of experience writing python scripts through my degree, but none using github or with a project of this scope so it's a big learning experience for me. 

I'm not sure how much I'll be able to contribute over the summer, but I'd like to do as much as possible and see this through :)

**Reply by JOJ0** [reply]
Guys, aren't we mixing up things here? The initial description in this issue was about replacing _before_ import (to improve matching likeliness).

The feature idea looks great @benpage2 but please dont get me wrong, this is confusing. Please open a new issue with that feature proposal.

**Reply by arsaboo** [reply]
@JOJ0 can we not include that feature in this framework? It still is a replacement...may be we add another config.

**Reply by benpage2** [reply]
@JOJ0 I just noticed you were specifically talking about adding regex to `fromfilename`. 

I'm still new to github so sorry if this is an obvious question, but if I were to open a new thread, where would be the best place to put it?

Thanks for all the help guys!

**Reply by JOJ0** [reply]
fromfilename also acts before import, your feature proposal sounds like it's ment to run on already existing library items.

But actually I'd like to see any solution in beets to prep super broken metadata to get a chance of being detected. Furthermore the beatport (Original Mix) dilemma could be fixed with such a pre-import-fix.

Pls clarify if I'm mistaken.

I would obly like to keep issues in order as good as possible :-)

**Reply by JOJ0** [reply]
> @JOJ0 I just noticed you were specifically talking about adding regex to `fromfilename`. 
> 
> 
> 
> I'm still new to github so sorry if this is an obvious question, but if I were to open a new thread, where would be the best place to put it?
> 
> 
> 
Simply open a new _issue_. I'm on mobile since travelling currently, not sure how to describe the clicks on Desktop. Youll figure it out. Cooy over your proposal into that new issue and let's refine it over there.

**Reply by JOJ0** [reply]
> > 
> > 
> > I'm still new to github so sorry if this is an obvious question, but if I were to open a new thread, where would be the best place to put it?
> > 
> > 
> > 
Simply open a new _issue_. I'm on mobile since travelling currently, not sure how to describe the clicks on Desktop. Youll figure it out. Copy over your proposal into that new issue and let's refine it over there.

**Comment by benpage2** [comment]
Yeah exactly, the idea is that you could do it on already imported items and potentially before import as well. Should the new issue go under ideas or feature requests?  @JOJ0 

Edit: I'll just throw it under ideas.
